### ISSUE TITLE
Using comment-based help makes dynamic parameters and validation-set values disappear

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#6694]
Adding comment-based help to an advanced function unexpectedly has the following impact on `Get-Help` / `-?`:

* static parameters with `[ValidateSet()]` attributes no longer show the valid values.

* dynamic parameters disappear completely (both from the syntax diagram and when using `Get-Help -Parameter`)

Steps to reproduce
------------------

Note: For brevity, the following repro _combines_ the two triggering conditions into a single function (static parameter + validation set vs. dynamic parameters); however, the problem occurs for each condition _in isolation_ too.

```powershell
using namespace System.Management.Automation

# A function with a static parameter constrained by a set of valid values and a similar dynamic parameter.
function Get-Foo1 {
  [CmdletBinding()]
  param(
    [ValidateSet('one', 'two')]
    [string] $Bar
  )

  DynamicParam {
    $paramAttr = [ParameterAttribute]::new()
    $validateSetAttr = [ValidateSetAttribute]::new([string[]] ('three', 'four'))
    $dynParam1 = [RuntimeDefinedParameter]::new('Baz', [string], ($paramAttr, $validateSetAttr))
    $paramDict = [RuntimeDefinedParameterDictionary]::new()
    $paramDict.Add($dynParam1.Name, $dynParam1)    
    $paramDict
  }

}

# Identical function, but with (minimal) comment-based help.
function Get-Foo2 {
<#
.SYNOPSIS
hi
#>
  [CmdletBinding()]
  param(
    [ValidateSet('one', 'two')]
    [string] $Bar
  )

  DynamicParam {
    $paramAttr = [ParameterAttribute]::new()
    $validateSetAttr = [ValidateSetAttribute]::new([string[]] ('three', 'four'))
    $dynParam1 = [RuntimeDefinedParameter]::new('Baz', [string], ($paramAttr, $validateSetAttr))
    $paramDict = [RuntimeDefinedParameterDictionary]::new()
    $paramDict.Add($dynParam1.Name, $dynParam1)    
    $paramDict
  }

}

# Invoke help for both functions:
Get-Foo1 -? | out-string
'----------------------------------------------------------'
Get-Foo2 -? | Out-String
```

Expected behavior
-----------------

```none
[...]
Get-Foo1 [[-Bar] {one | two}] [-Baz {three | four}]  [<CommonParameters>]
[...]
----------------------------------------------------------
[...]
Get-Foo1 [[-Bar] {one | two}] [-Baz {three | four}]  [<CommonParameters>]
[...]
```

Actual behavior
---------------

```none
[...]
    Get-Foo1 [[-Bar] {one | two}] [-Baz {three | four}]  [<CommonParameters>]
[...]
----------------------------------------------------------
[...]
    Get-Foo2 [[-Bar] <String>] [<CommonParameters>]
[...]
```

Note how the addition of comment-based help:

* made `-Baz` disappear

* devolved `-Bar` to showing just its type, not the valid values.

Environment data
----------------

```powershell
PowerShell Core v6.0.2 on macOS 10.13.4
PowerShell Core v6.0.2 on Ubuntu 16.04.4 LTS
PowerShell Core v6.0.2 on Microsoft Windows 10 Pro (64-bit; Version 1709, OS Build: 16299.371)
Windows PowerShell v5.1.16299.251 on Microsoft Windows 10 Pro (64-bit; Version 1709, OS Build: 16299.371)
```

**Comment by endowdly** [comment]
I believe this is a known issue. Here is an [explanation](https://info.sapien.com/index.php/scripting/scripting-help/writing-help-for-dynamic-parameters)

**Comment by mklement0** [comment]
Thanks, @endowdly: while the linked issue is definitely related, it is distinct:

_This_ issue is about the _syntax diagram devolving_, simply by the inclusion of comment-based help.

The linked issue is about the inability to _document dynamic parameters_ using comment-based help.

Can I ask you to create a _new_ issue in this repo in which you summarize the linked article?
You can ping @juneb, the author of the linked article.

**Comment by endowdly** [comment]
@mklement0 I see! I misunderstand the exact issue. Since @juneb discusses a _hack_ that was implemented via MAML to fix the inability to document dynamic parameters, I am hesitant to open a new issue regarding that. 

This issue, syntax diagram devolution, and the inability for `Get-Help` to read Comment-based help seem to be linked. The root issue seems to be the way Get-Help operates.

**Comment by johlju** [comment]
This problem is happening on parameters without using dynamic parameters too. Not opening a separate issue since this issue probably resolves this too.

```powershell
function Test1
{
    param
    (
        [ValidateSet('Value1', 'Value2')]
        $SecurableClass
    )
}

<#
    .SYNOPSIS
        Will not return the validate set from the parameter due to the
        comment-based help was added.
#>
function Test2
{
    param
    (
        [ValidateSet('Value1', 'Value2')]
        $SecurableClass
    )
}
```

Output

```
PS > Get-Help Test1

NAME
    Test1

SYNTAX
    Test1 [[-SecurableClass] {Value1 | Value2}]

ALIASES
    None

REMARKS
    None
```
```
PS > Get-Help Test2

NAME
    Test2

SYNOPSIS
    Will not return the validate set from the parameter due to the
    comment-based help was added.

SYNTAX
    Test2 [[-SecurableClass] <Object>] [<CommonParameters>]

DESCRIPTION

RELATED LINKS

REMARKS
    To see the examples, type: "get-help Test2 -examples".
    For more information, type: "get-help Test2 -detailed".
    For technical information, type: "get-help Test2 -full".
```

**Comment by yecril71pl** [comment]
I think ``Get-Help`` should look at parameter attributes instead of peeking into scripts.  This is currently required by Pester (or any other framework using run-time plug-ins).

**Comment by jmarrec** [comment]
I just found out that I have the same issue. Any plans to fix it please?

**Comment by ThomasNieto** [comment]
> This problem is happening on parameters without using dynamic parameters too. Not opening a separate issue since this issue probably resolves this too.
> 
> ```powershell
> function Test1
> {
>     param
>     (
>         [ValidateSet('Value1', 'Value2')]
>         $SecurableClass
>     )
> }
> 
> <#
>     .SYNOPSIS
>         Will not return the validate set from the parameter due to the
>         comment-based help was added.
> #>
> function Test2
> {
>     param
>     (
>         [ValidateSet('Value1', 'Value2')]
>         $SecurableClass
>     )
> }
> ```
> 
> Output
> 
> ```
> PS > Get-Help Test1
> 
> NAME
>     Test1
> 
> SYNTAX
>     Test1 [[-SecurableClass] {Value1 | Value2}]
> 
> ALIASES
>     None
> 
> REMARKS
>     None
> ```
> 
> ```
> PS > Get-Help Test2
> 
> NAME
>     Test2
> 
> SYNOPSIS
>     Will not return the validate set from the parameter due to the
>     comment-based help was added.
> 
> SYNTAX
>     Test2 [[-SecurableClass] <Object>] [<CommonParameters>]
> 
> DESCRIPTION
> 
> RELATED LINKS
> 
> REMARKS
>     To see the examples, type: "get-help Test2 -examples".
>     For more information, type: "get-help Test2 -detailed".
>     For technical information, type: "get-help Test2 -full".
> ```

This particular problem has been resolved and works on 7.2.4 but the original issue is still present.
